File: components/MachineLearning/FeatureExtraction.js
import { Audio } from "expo-av";

function resizeArray(arr, shape) {
  const resizedArray = [];

  for (let i = 0; i < shape[0]; i++) {
    const row = [];

    for (let j = 0; j < shape[1]; j++) {
      row.push(arr[i][j % arr[i].length]);
    }

    resizedArray.push(row);
  }

  return resizedArray;
}

function getMean(arr) {
  const meanArray = [];

  for (let i = 0; i < arr.length; i++) {
    const mean = arr[i].reduce((a, b) => a + b, 0) / arr[i].length;
    meanArray.push(mean);
  }

  return meanArray;
}

export default async function extractFeature(
  file_path,
  mfcc,
  chroma,
  mel,
  tonnetz
) {
  try {
    // Load the audio file using Expo's Audio API
    const soundObject = new Audio.Sound();
    await soundObject.loadAsync({ uri: file_path });
    const status = await soundObject.getStatusAsync();

    const { durationMillis, sampleRate } = status;

    const X = durationMillis;
    const sample_rate = sampleRate;

    console.log(41, file_path, status);
    let result = [];
    // lot of bugs
    // if (mfcc) {
    //   let mfccs = librosa.feature.mfcc(X, sample_rate, { n_mfcc: 40 });
    //   mfccs = resizeArray(mfccs, [mfccs.shape[0], fixed_length]);
    //   mfccs = getMean(mfccs.T);
    //   result = result.concat(mfccs);
    // }

    // if (chroma) {
    //   const stft = np.abs(librosa.stft(X));
    //   let chroma = librosa.feature.chroma_stft(stft, sample_rate);
    //   chroma = resizeArray(chroma, [chroma.shape[0], fixed_length]);
    //   chroma = getMean(chroma.T);
    //   result = result.concat(chroma);
    // }

    // if (mel) {
    //   let mel = librosa.feature.melspectrogram(X, sample_rate);
    //   mel = resizeArray(mel, [mel.shape[0], fixed_length]);
    //   mel = getMean(mel.T);
    //   result = result.concat(mel);
    // }

    // if (tonnetz) {
    //   let tonnetz = librosa.feature.tonnetz(X, sample_rate);
    //   tonnetz = resizeArray(tonnetz, [tonnetz.shape[0], fixed_length]);
    //   tonnetz = getMean(tonnetz.T);
    //   result = result.concat(tonnetz);
    // }

    // Use the 'result' array as needed
    console.log("Extracted features:");
    return result;
  } catch (error) {
    console.log("Error loading audio file:", error);
  }
}

File: backend/controllers/python-files/featureExtraction.py
import librosa
import numpy as np
import soundfile
from scipy import stats
from sklearn.preprocessing import normalize


def extract_feature(file_name, mfcc, chroma, mel, tonnetz):
    fixed_length = 1000
    try:
        with soundfile.SoundFile(file_name) as sound_file:
            X = sound_file.read(dtype="float32")
            sample_rate = sound_file.samplerate
            # print(sample_rate)

            result = np.array([])

            # if mfcc:
            #     mfccs=np.mean(librosa.feature.mfcc(y=X, sr=sample_rate, n_mfcc=40).T, axis=0)
            #     result=np.hstack((result, mfccs))
            if mfcc:
                mfccs = np.mean(librosa.feature.mfcc(
                    y=X, sr=sample_rate, n_mfcc=40).T, axis=0)
                # mfccs = np.pad(mfccs, (0, pad_length - len(mfccs)), mode='constant')
                result = np.hstack((result, mfccs))

            if chroma:
                stft = np.abs(librosa.stft(X))
                chroma = np.mean(librosa.feature.chroma_stft(
                    S=stft, sr=sample_rate).T, axis=0)
                result = np.hstack((result, chroma))

            if mel:
                # mel=np.mean(librosa.feature.melspectrogram(X, sr=sample_rate).T,axis=0)
                mel = np.mean(librosa.feature.melspectrogram(
                    y=X, sr=sample_rate).T, axis=0)

                result = np.hstack((result, mel))

            if tonnetz:
                ton = np.mean(librosa.feature.tonnetz(
                    X, sr=sample_rate).T, axis=0)
                result = np.hstack((result, ton))

        return result
    except Exception as e:
        print(f"Error extracting feature: {str(e)}")
        return None


def data_post_processing(data):
    s_sample = np.array(data)

    # Reshape s_sample to have the correct shape
    s_sample_reshaped = np.expand_dims(
        s_sample, axis=0)  # Add a batch dimension
    s_sample_reshaped = normalize(s_sample_reshaped, axis=1, norm='l1')
    print(s_sample_reshaped.min())
    print(s_sample_reshaped.max())
    return s_sample_reshaped


if __name__ == "__main__":
    import sys
    file_path = sys.argv[1]
    result = extract_feature(file_path, True, True, True, False)

    s_sample_reshaped = data_post_processing(result)

    print(s_sample_reshaped)
